### PR TITLE
[30007] Allow users to enable CORS on APIv3 resources

### PR DESCRIPTION
**Pull request by oliverguenther**
Allows a setting and configuration option to allow access to APIV3.

Setting allows us to selectively enable this for cloud instances too.

https://community.openproject.com/wp/30007

### CODE REVIEW

File: config/initializers/rack-cors.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See docs/COPYRIGHT.rdoc for more details.
+#++
+Rails.application.config.middleware.insert_before 0, Rack::Cors do
+  allow do
+    origins '*'
```
Comment (ulferts) [inline config/initializers/rack-cors.rb]: I think it better to not set the wildcard origin here as this limits the ability to provide credentials (https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSNotSupportingCredentials). E.g. we could alter the configuration/setting to take a value that is then passed to `origins`.

**Review by ulferts — CHANGES_REQUESTED** [review]
Almost complete, but we should change the `origins` value as explained below.

**Comment by oliverguenther** [comment]
I made the entire thing accessible by API settings in the UI and let users configure the origins they want to open up to. @machisuji @ulferts please take another look at this if possible. There's a nice tool at https://www.test-cors.org/ that allows you to test these configurations.

**Review by machisuji — DISMISSED** [review]
Looking good and working. Thanks! Just one thing:

Right now the "API V3 Cross-Origin Resource Sharing (CORS) allowed origins" are required for it to work. Saving without any origins (which is possible) doesn't actually make CORS work.

So either we make the field required or do the reflective thing and simply allow as origin what ever is in `request.host` during the OPTIONS request.

**Review by machisuji — APPROVED** [review]
:+1:

**Pull request merged**